cli/command/cli.go

```diff
@@ -313,13 +325,21 @@ func resolveDefaultDockerEndpoint(opts *cliflags.CommonOptions) (docker.Endpoint
 	}, nil
 }
 
+func (cli *DockerCli) getInitTimeout() time.Duration {
+	if cli.initTimeout != 0 {
+		return cli.initTimeout
+	}
+	return defaultInitTimeout
+}
+
 func (cli *DockerCli) initializeFromClient() {
 	ctx := context.Background()
-	if strings.HasPrefix(cli.DockerEndpoint().Host, "tcp://") {
+	host := cli.DockerEndpoint().Host
+	if strings.HasPrefix(host, "tcp://") || strings.HasPrefix(host, "unix://") {
 		// @FIXME context.WithTimeout doesn't work with connhelper / ssh connections
 		// time="2020-04-10T10:16:26Z" level=warning msg="commandConn.CloseWrite: commandconn: failed to wait: signal: killed"
 		var cancel func()
-		ctx, cancel = context.WithTimeout(ctx, 2*time.Second)
+		ctx, cancel = context.WithTimeout(ctx, cli.getInitTimeout())
```

Comment:
I'm curious though; https://github.com/docker/cli/issues/3652 describes that the CLI hangs indefinitely; but there's already a 2 second timeout. I understand that this PR makes that timeout configurable, but wouldn't https://github.com/docker/cli/issues/3652 already timeout after 2 seconds? Or is the `unix://` addition the actual fix here?

Reply:
the `unix://` addition is the fix. The configurability isn't the focus of this pr, it's something @vvoland asked for in review.

Reply:
Thanks for those details (I probably should've read back the whole discussion above)

So some thoughts;

First of all, "the fix"; I think this was an incorrect assumption on my side. I added this check in https://github.com/docker/cli/commit/b39739123b845f872549e91be184cc583f5b387c / https://github.com/docker/cli/pull/2424, wich the assumption that a local socket would not really need a timeout (unless something was horribly wrong), and of course "horribly wrong" now appears to not be theoretical (per your ticket) :joy:

I assume the same _could_ also happen on a named pipe (`npipe://`) on Windows or `fd://` (default on Linux for systemd created socket)? If so, we could be explicit;

```go
if !strings.HasPrefix(cli.DockerEndpoint().Host, "ssh://") {
```

We should verify though if `cli.DockerEndpoint().Host` is guaranteed to always have a scheme (I know there's some "fuzzy" / "relaxed" handling of some values in our code-base, and we should verify that the values is at least normalized (scheme added if missing) early on). At a quick glance, I think that's the case though; calls `getServerHost()`, which does normalising (lol, we need to get rid of some abstraction here and there); https://github.com/docker/cli/blob/a32cd16160f1b41c1c4ae7bee4dac929d1484e59/cli/command/cli.go#L324

As to making the timeout configurable; in _general_, I'm not against adding configurable options; that said, once added, it's pretty hard to (deprecate and) remove an option, so I also try to be conservative where possible. I realize the 2 seconds was just some arbitrary value ("2 seconds should be enough?"); for your use-case, are you planning on adjusting that default, or would the "fix" be enough (only needed to prevent a full hang, but default is "fine" otherwise)?

What I'm thinking is, that we could; split the fix from the feature; the fix (one line change, and appears to be low risk) could be candidate for backporting to 20.10 if we think that's good. For the feature (make the timeout configurable), we could keep that in draft (for now), and still decide to take it if it turns out there's good use-cases for overriding. Ultimately (also see my other comment above), I think it would be great if we didn't bother with pinging at all, _unless_ we need it (in most cases that would be "the moment we need to connect to the API anyway").